### PR TITLE
Prepare for whole stage retries

### PR DESCRIPTION
**Pull request by arhimondr**
Introduce stage execution attempt concept. Extend TaskId with the stage execution id 

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by arhimondr** [comment]
@wenleix @rschlussel Although we discussed something more like `AttemptId`, i decided to go with something like `StageExecutionId`.

The reason for that is that currently

- Stage has `StageExecution`
- Attempt it is not someting we create, but something we perform. In other words we perform an attempt by creating a new instance of `StageExectuion` which then have an unique identifier, `StageExecutionId`.

Then the `StageInfo` has two fields

```
 private final StageExecutionInfo latestAttemptExecutionInfo;
 private final List<StageExecutionInfo> previousAttemptsExecutionInfos;
```

Please let me know what you think

**Comment by arhimondr** [comment]
P.S.: Still WIP and probably not ready for review. I'm still fixing UI, and have no idea whether the tests will pass =)

**Comment by arhimondr** [comment]
@wenleix @rschlussel Ready for review

**Comment by arhimondr** [comment]
@rschlussel Could you please link this PR to the issue, once you create one?

**Comment by arhimondr** [comment]
UI: 

Selection_138

**Comment by wenleix** [comment]
@arhimondr : How does the UI already be able to show failed stage? -- whole stage retry is not supported yet right? 😮

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 1]
```diff
@@ -951,18 +944,9 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
                         }
                     })
                     .collect(toImmutableSet());
-
-            this.childStageIds = children.stream()
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: interesting. So previously this part of code is not useful?

**Comment by arhimondr** [comment]
> So this basically moves "location, fragment, session, renderedPlan"

Basically this is a preparation for the `Prepare StageInfo for whole stage retries`. In the `Prepare StageInfo for whole stage retries` commit the `StageExecutionStateMachine` no longer creates `StageInfo`. It creates the `StageExecutionInfo` instead. The `StageExecutionInfo` doesn't include the `PlanFragment`/`Location`, and the `Session` parameter also becomes extra. 

the `renderedPlan` had been already unused before i removed it.

**Comment by arhimondr** [comment]
> Also what's this location for?

That was my question as well. Turns out it is needed to get the Stage URL for partial cancel.

**Comment by arhimondr** [comment]
> How does the UI already be able to show failed stage? -- whole stage retry is not supported yet right? open_mouth

I just hardcoded failed executions to get this screenshot =))

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java (hunk 1 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: Yeah, it was a dead code

File: presto-main/src/main/java/com/facebook/presto/event/SplitMonitor.java [hunk 2]
```diff
@@ -77,7 +77,8 @@ private void splitCompletedEvent(TaskId taskId, DriverStats driverStats, @Nullab
             eventListenerManager.splitCompleted(
                     new SplitCompletedEvent(
                             taskId.getQueryId().toString(),
-                            taskId.getStageId().toString(),
+                            taskId.getStageExecutionId().getStageId().toString(),
+                            taskId.getStageExecutionId().toString(),
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/event/SplitMonitor.java]: I was just trying to mimic the way it was before (not sure if that's the best way). So before it used to inlude queryId + full stageId (that also includes query id). Now there will be query id + full stage id (that includes query id) + full stage execution id (that also includes stage id and query id)

File: presto-main/src/test/java/com/facebook/presto/execution/TestQueryStats.java [hunk 3]
```diff
@@ -201,6 +201,7 @@
 
             ImmutableList.of(new StageGcStatistics(
                     101,
+                    1001,
```
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/execution/TestQueryStats.java]: retried 1000 times? 😨

**Pull request merged**